### PR TITLE
Add static-link support option

### PR DESCRIPTION
**Pull request by Rob1in**
Manually tested on:

- linux arm64
- linux amd64
- Darwin arm64
- Darwin amd64

Also tested the integration in rdk on MacOS (one main sending encoded jpegs to a remote which decodes them).

### CODE REVIEW

File: jpeg/include/jmorecfg.h [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+  -1
+};
+
+static const int rgb_pixelsize[JPEG_NUMCS] = {
+  -1, -1, RGB_PIXELSIZE, -1, -1, -1, EXT_RGB_PIXELSIZE, EXT_RGBX_PIXELSIZE,
+  EXT_BGR_PIXELSIZE, EXT_BGRX_PIXELSIZE, EXT_XBGR_PIXELSIZE, EXT_XRGB_PIXELSIZE,
+  EXT_RGBX_PIXELSIZE, EXT_BGRX_PIXELSIZE, EXT_XBGR_PIXELSIZE, EXT_XRGB_PIXELSIZE,
+  -1
+};
+
```
Comment (edaniels) [inline jpeg/include/jmorecfg.h:356]: we need to include the license for the libjpeg project(s) somewhere

File: jpeg/jpeg_static.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+Contents of lib/ have been built from that source as well.
+Libraries are actually copies of libturbojpeg.a as built for each architecture.
+*/
+
+/*
+#cgo CFLAGS: -I${SRCDIR}/include
+#cgo linux,amd64 LDFLAGS: ${SRCDIR}/lib/libjpeg_linux_amd64.a
+#cgo linux,arm64 LDFLAGS: ${SRCDIR}/lib/libjpeg_linux_arm64.a
+#cgo darwin,amd64 LDFLAGS: ${SRCDIR}/lib/libjpeg_darwin_amd64.a
+#cgo darwin,arm64 LDFLAGS: ${SRCDIR}/lib/libjpeg_darwin_arm64.a
```
Comment (edaniels) [inline jpeg/jpeg_static.go]: are we explicitly not supporting windows which may now fail since the default is not dynamic

**Review by edaniels — CHANGES_REQUESTED** [review]
Couple of things in comments. Major thing though is documenting how to build these static libraries for future changes.

Also, @Otterverse why do we even need static libs over just cgo compiling this code on demand since it's pretty self contained?

**Comment by Otterverse** [comment]
> Couple of things in comments. Major thing though is documenting how to build these static libraries for future changes.

So originally as a PoC I just pulled these out of the homebrew prebuilt packages. But talking with Robin about windows, realized there are official packages here: https://sourceforge.net/projects/libjpeg-turbo/files/2.1.5.1/ I'd suggest @Rob1in extract the files from there (replacing the ones I put into this PR as the PoC) and document the process.

> Also, @Otterverse why do we even need static libs over just cgo compiling this code on demand since it's pretty self contained?

Not sure I follow there. You want to include the entire source code for libjpeg and compile it on demand every time? I'm not sure that would work cleanly as it's a complex build process AFAIK (uses cmake and such.) Or did you mean something else?

**Comment by Rob1in** [comment]
@edaniels @Otterverse 
 I have added:

-  windows static library (and specific headers)
- building instructions (see jpeg/BUILDING.md)
- license (see jpeg/LICENSE.md)

Tested on windows.

File: jpeg/BUILDING.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# Instructions to build the static libraries
+
+## Current libraries version
+
+The current static libraries are version 2.1.5.1 (see [official binaries](https://sourceforge.net/projects/libjpeg-turbo/files/2.1.5.1/) and [Github Project Page](https://github.com/libjpeg-turbo/libjpeg-turbo)).
+## Darwin and Linux
+The static libraries are extracted from the brew bottles.  
+To do so, you need to have [homebrew](https://brew.sh) installed.
+
+1. Install , libjpeg-turbo : `brew install jpeg-turbo`
```
Comment (Otterverse) [inline jpeg/BUILDING.md]: You don't have to install it to extract the bottles.

File: jpeg/BUILDING.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Instructions to build the static libraries
+
+## Current libraries version
+
+The current static libraries are version 2.1.5.1 (see [official binaries](https://sourceforge.net/projects/libjpeg-turbo/files/2.1.5.1/) and [Github Project Page](https://github.com/libjpeg-turbo/libjpeg-turbo)).
+## Darwin and Linux
```
Comment (Otterverse) [inline jpeg/BUILDING.md]: As this only works for x86_64 linux (there is now brew for arm64 linux), I would suggest changing instructions to use the official builds for linux AND windows, and only use brew for mac.) Assuming that works.

File: jpeg/BUILDING.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+The static libraries are extracted from the brew bottles.  
+To do so, you need to have [homebrew](https://brew.sh) installed.
+
+1. Install , libjpeg-turbo : `brew install jpeg-turbo`
+2. Install jq `brew install jq` (not crucial but easier to process the json)
+3. run `brew info --json jpeg-turbo | jq -r ".[].bottle"` to display the URLs for the bottles of the stable releases are.
+4. Choose the architecture and the OS you want to build the static library for and copy its URL.
```
Comment (Otterverse) [inline jpeg/BUILDING.md]:
```suggestion
4. Choose the architecture and the OS you want to extract the static library for and copy its URL.
```

File: jpeg/BUILDING.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+5. Get the bottle `curl -L -H "Authorization: Bearer QQ==" -o MyFilename.tar.gz $URL` (replace the $URL field)
+6. Extract the libraries and the header files with `tar -xf MyFilename.tar.gz `
+
```
Comment (Otterverse) [inline jpeg/BUILDING.md:18]: Need more steps here.
7. Find libturbojpeg.a and copy/rename it for the target architectured: ex "libturbojpeg_linux_arm64.a" This is the ONLY file you need for each architecture.
8. On one architecture only (it doesn't matter which) you must also extract all the .h files from the include folder.

File: jpeg/include/jconfig.h
Comment (Otterverse) [inline jpeg/include/jconfig.h:1]: As this is the only file that differs between windows and the rest, can you confirm that it doesn't different between architectures either? It's setting the jpeg_lib_version to 80 (when windows is 62) and it's also setting some support flags. Need to make sure linux/mac and arm/x86 all use the same version of this if we only include the one.

File: jpeg/jpeg_static.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+package jpeg
+
+/*
+Note the contents of include/ are extracted from jpeg-turbo 2.1.5.1.
+Contents of lib/ have been built from that source as well.
+Libraries are actually copies of libturbojpeg.a as built for each architecture.
+*/
+
+/*
+#cgo linux, darwin CFLAGS: -I${SRCDIR}/include
```
Comment (Otterverse) [inline jpeg/jpeg_static.go]:
```suggestion
#cgo linux,darwin CFLAGS: -I${SRCDIR}/include
```
Can't have spaces here.

File: jpeg/LICENSE.md
Comment (Otterverse) [inline jpeg/LICENSE.md:1]: I'm not a lawyer, but I don't think you can just include the jpeg-turbo license as our own license. As far as I understand it, you need to follow the instructions IN this file about how to comply.

**Review by Otterverse — CHANGES_REQUESTED** [review]
Need some updates to the build doc and license stuff, and I'd like confirmation about headers matching across unixes. At a technical level looks pretty good. Just one bug I found.

File: jpeg/LICENSE.md
Comment (edaniels) [reply jpeg/LICENSE.md:1]: Whoops, that was my bad advice to just simply include this. Sorry @Rob1in

File: jpeg/BUILDING.md (hunk 3 above)
Comment (Rob1in) [reply jpeg/BUILDING.md]: You are right! I thought it was needed to run brew info. Removing this...

File: jpeg/include/jconfig.h
Comment (Rob1in) [reply jpeg/include/jconfig.h:1]: linux/mac and arm/x86 all use the same version of libjpeg (v8). It is set in the brew formula [here](https://github.com/Homebrew/homebrew-core/blob/e7bbc68273847775d450f47a5d69da09c961c2cd/Formula/jpeg-turbo.rb#L41) with -DWITH_JPEG8=1

File: jpeg/jpeg_static.go (hunk 7 above)
Comment (Rob1in) [reply jpeg/jpeg_static.go]: ok!  changing this

File: jpeg/BUILDING.md (hunk 4 above)
Comment (Rob1in) [reply jpeg/BUILDING.md]: Extracting the libraries and header files from the official binaries doesn't work on linux x86_64. I haven't tested it on arm64 linux (but works when extracted from the brew bottles). So I would just add a link to the brew jpeg-turbo URL on this introductory paragraph if it's ok?

Comment (Otterverse) [reply jpeg/BUILDING.md]: Okay, if the official ones don't work, that's fine then. It's just a shame there's now brew bottle for linux/arm64 and thus someone has to build it manually. Oh well, at least these shouldn't need to update very often.

File: jpeg/LICENSE.md
Comment (Rob1in) [reply jpeg/LICENSE.md:1]: Changed the license to comply with [this](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/LICENSE.md?plain=1#L55).

File: jpeg/BUILDING.md (hunk 6 above)
Comment (Rob1in) [reply jpeg/BUILDING.md:18]: done

File: jpeg/BUILDING.md (hunk 5 above)
Comment (Rob1in) [reply jpeg/BUILDING.md]: changed it!

**Review by Otterverse — APPROVED** [review]
LGTM, with one big note: This is a PR to merge into the branch called "fork" which means most people won't be able to `go get ...` this without extra instructions. Are there plans to merge this into main afterwards? If not, may need to modify the instructions and such. 

Personally, I'd rename the "main" branch "upstream" and make this modified version the "main" branch based ON that "upstream" branch, instead of the other way around. But as I think that's out of the scope of the PR/code itself, I'll leave that to you for later.

File: jpeg/include/jmorecfg.h (hunk 1 above)
Comment (Rob1in) [reply jpeg/include/jmorecfg.h:356]: done

File: jpeg/jpeg_static.go (hunk 2 above)
Comment (Rob1in) [reply jpeg/jpeg_static.go]: added the static library for windows

**Pull request merged**